### DISCUSSION TITLE
How to compile single threaded filament?

### DISCUSSION THREAD

**Discussion by FinleyPan** [google/filament#4540] [Q&A]
Hey, I found a macro `FILAMENT_SINGLE_THREADED` in `compiler.h` can be defined to enable single thread:
```cpp
#if defined(__EMSCRIPTEN__) || defined(FILAMENT_SINGLE_THREADED)
#   define UTILS_HAS_THREADING 0
#else
#   define UTILS_HAS_THREADING 1
#endif
```
But I have to add this macro in the Xcode project after it generated by cmake. My question is whether I can directly build single threaded filament by `build.sh`?

**Comment by romainguy** [comment]
You could invoke CMake yourself instead of using build.sh. Note that this option only exists for the WebAssembly target and as soon as we reasonably can we will probably remove it. There should be no reason to use it for other targets.

**Reply by FinleyPan** [reply]
OK. BTW I think it is useful for that I need call OpenGL API outside of filament and scheduling these calls in the same thread might be an easier way to go.

**Reply by romainguy** [reply]
You can use a shared context for this. There's no guarantee we'll keep supporting OpenGL and doing what you want to do can lead to annoying bugs since filament does state tracking internally to minimize driver calls.

**Reply by heitaoflower** [reply]
I think there are a few practical reasons that suggest filament keep this single-thread running mode or feature.

For example, If at some media types applications, the external environment needs to maintain an EGLContext,  at the same time, need filament to do some filter stuff, if we only depend on the multi-thread-rendering feature, that meant we need to wait for the job completion of filament.  that means we must tolerate two times glFinish() at least. no matter online or offline.
The lag time is marvelous!.

however, the top or external application needs to do some messy sync-up operation, to guarantee the pure OpenGL code and the filament context code work very well.

Personally speaking, that is not kindly, besides the entire application based on filament thread.

so the better solution in my eyes is to provide two running strategies, one is immediate rendering mode, the other one is delay rendering mode.
